File: openframe/services/openframe-api/src/main/java/com/openframe/api/service/ApiKeyService.java
package com.openframe.api.service;

import com.openframe.api.dto.ApiKeyResponse;
import com.openframe.api.dto.CreateApiKeyRequest;
import com.openframe.api.dto.CreateApiKeyResponse;
import com.openframe.api.dto.UpdateApiKeyRequest;
import com.openframe.core.exception.ApiKeyException;
import com.openframe.core.model.ApiKey;
import com.openframe.data.model.mongo.ApiKeyStatsMongo;
import com.openframe.data.repository.mongo.ApiKeyRepository;
import com.openframe.data.repository.mongo.ApiKeyStatsMongoRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.security.SecureRandom;
import java.time.Instant;
import java.time.ZoneOffset;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class ApiKeyService {
    
    private final ApiKeyRepository apiKeyRepository;
    private final ApiKeyStatsMongoRepository apiKeyStatsMongoRepository;
    private final PasswordEncoder passwordEncoder;
    private final SecureRandom secureRandom = new SecureRandom();
    
    private static final String KEY_ID_PREFIX = "ak_";
    private static final String SECRET_PREFIX = "sk_";
    private static final int KEY_ID_LENGTH = 16;
    private static final int SECRET_LENGTH = 32;
    
    /**
     * Create a new API key for a user
     */
    public CreateApiKeyResponse createApiKey(String userId, CreateApiKeyRequest request) {
        log.info("Creating API key '{}' for user: {}", request.name(), userId);
        
        try {
            return buildAndSaveApiKey(userId, request.name(), request.description(), request.expiresAt());
        } catch (Exception e) {
            log.error("Failed to create API key for user: {}", userId, e);
            throw new ApiKeyException("Failed to create API key", e);
        }
    }

    /**
     * Get all API keys for a user
     */
    public List<ApiKeyResponse> getApiKeysForUser(String userId) {
        return apiKeyRepository.findByUserId(userId).stream()
            .map(this::mapToResponse)
            .toList();
    }
    
    /**
     * Get API key by ID for a specific user
     */
    public ApiKeyResponse getApiKeyById(String keyId, String userId) {
        ApiKey apiKey = apiKeyRepository.findByIdAndUserIdOrElseThrow(keyId, userId);
        return mapToResponse(apiKey);
    }
    
    /**
     * Update API key
     */
    public ApiKeyResponse updateApiKey(String keyId, String userId, UpdateApiKeyRequest request) {
        ApiKey apiKey = apiKeyRepository.findByIdAndUserIdOrElseThrow(keyId, userId);
        
        if (request.name() != null) {
            apiKey.setName(request.name());
        }
        if (request.description() != null) {
            apiKey.setDescription(request.description());
        }
        if (request.enabled() != null) {
            apiKey.setEnabled(request.enabled());
        }
        if (request.expiresAt() != null) {
            apiKey.setExpiresAt(request.expiresAt());
        }
        
        apiKey.setUpdatedAt(Instant.now());
        
        return mapToResponse(apiKeyRepository.save(apiKey));
    }
    
    /**
     * Delete API key and its statistics
     */
    public void deleteApiKey(String keyId, String userId) {
        ApiKey apiKey = apiKeyRepository.findByIdAndUserIdOrElseThrow(keyId, userId);

        apiKeyRepository.delete(apiKey);
        apiKeyStatsMongoRepository.deleteById(keyId);

        log.info("Deleted API key and its statistics: {}", keyId);
    }
    
    /**
     * Disable expired API keys TODO (scheduled task)
     */
    public void disableExpiredKeys() {
        List<ApiKey> expiredKeys = apiKeyRepository.findExpiredKeys(Instant.now());
        
        for (ApiKey key : expiredKeys) {
            key.setEnabled(false);
            key.setUpdatedAt(Instant.now());
            apiKeyRepository.save(key);
            log.info("Disabled expired API key: {}", key.getKeyId());
        }
    }
    

    private String generateKeyId() {
        byte[] bytes = new byte[KEY_ID_LENGTH / 2];
        secureRandom.nextBytes(bytes);
        return KEY_ID_PREFIX + bytesToHex(bytes);
    }
    
    private String generateSecret() {
        byte[] bytes = new byte[SECRET_LENGTH / 2];
        secureRandom.nextBytes(bytes);
        return bytesToHex(bytes);
    }
    
    private String bytesToHex(byte[] bytes) {
        StringBuilder result = new StringBuilder();
        for (byte b : bytes) {
            result.append(String.format("%02x", b));
        }
        return result.toString();
    }

    private ApiKeyResponse mapToResponse(ApiKey apiKey) {
        log.debug("Getting stats for API key: {}", apiKey.getKeyId());
        ApiKeyStatsMongo totalStats = apiKeyStatsMongoRepository.findById(apiKey.getKeyId()).orElse(null);

        if (totalStats != null) {
            log.debug("Found stats for {}: total={}, successful={}, failed={}",
                    apiKey.getKeyId(), totalStats.getTotalRequests(),
                    totalStats.getSuccessfulRequests(), totalStats.getFailedRequests());
        } else {
            log.debug("No stats found for API key: {}", apiKey.getKeyId());
        }

        return ApiKeyResponse.builder()
                .id(apiKey.getKeyId())
                .name(apiKey.getName())
                .description(apiKey.getDescription())
                .enabled(apiKey.isEnabled())
                .createdAt(apiKey.getCreatedAt())
                .updatedAt(apiKey.getUpdatedAt())
                .expiresAt(apiKey.getExpiresAt())
                .lastUsed(totalStats != null && totalStats.getLastUsed() != null ?
                        totalStats.getLastUsed().toInstant(ZoneOffset.UTC) : null)
                .totalRequests(totalStats != null ? totalStats.getTotalRequests() : 0L)
                .successfulRequests(totalStats != null ? totalStats.getSuccessfulRequests() : 0L)
                .failedRequests(totalStats != null ? totalStats.getFailedRequests() : 0L)
                .build();
    }

    /**
     * Regenerate API key (creates new key with same metadata but new secret)
     */
    public CreateApiKeyResponse regenerateApiKey(String keyId, String userId) {
        log.info("Regenerating API key {} for user: {}", keyId, userId);

        ApiKey existingKey = apiKeyRepository.findByIdAndUserIdOrElseThrow(keyId, userId);

        String name = existingKey.getName();
        String description = existingKey.getDescription();
        Instant expiresAt = existingKey.getExpiresAt();

        apiKeyRepository.delete(existingKey);
        log.debug("Deleted old API key: {}", keyId);

        return buildAndSaveApiKey(userId, name, description, expiresAt);
    }

    /**
     * Build and save API key with generated credentials
     */
    private CreateApiKeyResponse buildAndSaveApiKey(String userId, String name, String description, Instant expiresAt) {
        String keyId = generateKeyId();
        String secret = generateSecret();
        String hashedSecret = passwordEncoder.encode(secret);
        
        log.debug("Generated keyId: {}", keyId);
        log.debug("Generated secret length: {}", secret.length());

        ApiKey apiKey = ApiKey.builder()
            .keyId(keyId)
            .hashedKey(hashedSecret)
            .name(name)
            .description(description)
            .userId(userId)
            .enabled(true)
            .expiresAt(expiresAt)
            .createdAt(Instant.now())
            .updatedAt(Instant.now())
            .build();
        
        log.debug("Built API key with keyId: {}", apiKey.getKeyId());

        ApiKey savedKey = apiKeyRepository.save(apiKey);
        log.info("Saved API key with ID: {}", savedKey.getKeyId());

        return new CreateApiKeyResponse(
            mapToResponse(savedKey),
            keyId + "." + SECRET_PREFIX + secret
        );
    }
} 

File: openframe/services/openframe-ui/src/services/ApiKeyService.ts
import { restClient } from '../apollo/apolloClient';

export interface ApiKey {
  id: string;
  name: string;
  description?: string;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
  expiresAt?: string;
  lastUsed?: string;
  totalRequests?: number;
  successfulRequests?: number;
  failedRequests?: number;
}

export interface CreateApiKeyRequest {
  name: string;
  description?: string;
  expiresAt?: string;
}

export interface UpdateApiKeyRequest {
  name?: string;
  description?: string;
  enabled?: boolean;
  expiresAt?: string;
}

export interface CreateApiKeyResponse {
  apiKey: ApiKey;
  fullKey: string;
}

export interface ApiKeyUsageStats {
  keyId: string;
  totalRequests: number;
  requestsToday: number;
  requestsThisWeek: number;
  requestsThisMonth: number;
  lastUsedAt?: string;
  rateLimitExceeded: number;
}

export class ApiKeyService {
  private static readonly BASE_URL = '/api-keys';

  /**
   * Get all API keys for the current user
   */
  static async getApiKeys(): Promise<ApiKey[]> {
    try {
      const response = await restClient.get<ApiKey[]>(`${import.meta.env.VITE_API_URL}${this.BASE_URL}`);
      return response;
    } catch (error) {
      console.error('Failed to fetch API keys:', error);
      throw error;
    }
  }

  /**
   * Get a specific API key by ID
   */
  static async getApiKey(id: string): Promise<ApiKey> {
    try {
      const response = await restClient.get<ApiKey>(`${import.meta.env.VITE_API_URL}${this.BASE_URL}/${id}`);
      return response;
    } catch (error) {
      console.error(`Failed to fetch API key ${id}:`, error);
      throw error;
    }
  }

  /**
   * Create a new API key
   */
  static async createApiKey(request: CreateApiKeyRequest): Promise<CreateApiKeyResponse> {
    try {
      const response = await restClient.post<CreateApiKeyResponse>(
        `${import.meta.env.VITE_API_URL}${this.BASE_URL}`,
        request
      );
      return response;
    } catch (error) {
      console.error('Failed to create API key:', error);
      throw error;
    }
  }

  /**
   * Update an existing API key
   */
  static async updateApiKey(id: string, request: UpdateApiKeyRequest): Promise<ApiKey> {
    try {
      const response = await restClient.put<ApiKey>(
        `${import.meta.env.VITE_API_URL}${this.BASE_URL}/${id}`,
        request
      );
      return response;
    } catch (error) {
      console.error(`Failed to update API key ${id}:`, error);
      throw error;
    }
  }

  /**
   * Delete an API key
   */
  static async deleteApiKey(id: string): Promise<void> {
    try {
      await restClient.delete(`${import.meta.env.VITE_API_URL}${this.BASE_URL}/${id}`);
    } catch (error) {
      console.error(`Failed to delete API key ${id}:`, error);
      throw error;
    }
  }

  /**
   * Enable or disable an API key
   */
  static async toggleApiKey(id: string, enabled: boolean): Promise<ApiKey> {
    try {
      const response = await restClient.put<ApiKey>(
        `${import.meta.env.VITE_API_URL}${this.BASE_URL}/${id}`,
        { enabled }
      );
      return response;
    } catch (error) {
      console.error(`Failed to toggle API key ${id}:`, error);
      throw error;
    }
  }

  /**
   * Get usage statistics for an API key
   */
  static async getApiKeyUsage(keyId: string): Promise<ApiKeyUsageStats> {
    try {
      const response = await restClient.get<ApiKeyUsageStats>(
        `${import.meta.env.VITE_API_URL}${this.BASE_URL}/${keyId}/usage`
      );
      return response;
    } catch (error) {
      console.error(`Failed to fetch usage stats for API key ${keyId}:`, error);
      throw error;
    }
  }

  /**
   * Regenerate an API key (creates new secret)
   */
  static async regenerateApiKey(id: string): Promise<CreateApiKeyResponse> {
    try {
      const response = await restClient.post<CreateApiKeyResponse>(
        `${import.meta.env.VITE_API_URL}${this.BASE_URL}/${id}/regenerate`
      );
      return response;
    } catch (error) {
      console.error(`Failed to regenerate API key ${id}:`, error);
      throw error;
    }
  }

  /**
   * Format API key for display (show full key ID)
   */
  static formatKeyForDisplay(keyId: string): string {
    return keyId || '';
  }

  /**
   * Get status color for API key
   */
  static getStatusColor(apiKey: ApiKey): string {
    if (!apiKey.enabled) return 'red';
    if (apiKey.expiresAt && new Date(apiKey.expiresAt) < new Date()) return 'orange';
    return 'green';
  }

  /**
   * Get status text for API key
   */
  static getStatusText(apiKey: ApiKey): string {
    if (!apiKey.enabled) return 'Disabled';
    if (apiKey.expiresAt && new Date(apiKey.expiresAt) < new Date()) return 'Expired';
    return 'Active';
  }
} 